### ISSUE TITLE
Open Subprojects Automatically

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#529]
I would like to improve the project open/import logic so that to make it quicker to open all subprojects in a directory.  The primary use case for this is to make it easier to checkout repositories with multiple subprojects, like https://github.com/prominic/Moonshine-IDE.git , but I think we should apply this to the project open/import logic so that it can be used for local directories or workspaces as well.

Currently, the projects is like this:
1. Clone https://github.com/prominic/Moonshine-IDE.git
2. Decline to import the project
3. Open ide/MoonshineDESKTOPevolved
4. Open ide/MoonshineSharedCore
5. Open ide/MoonshineWEBevolved
6. ...

Instead I want it to be like this:
1. Clone https://github.com/prominic/Moonshine-IDE.git
2. Moonshine shows a list of the detected subprojects.  
3. The user chooses which of the projects to open.  By default, all are selected

The prompt would look something like this:
![image](https://user-images.githubusercontent.com/24532731/56436160-83fa0200-62a0-11e9-95d3-95d38d827185.png)
Mockup:  
[2019_04_19__Subproject_Import_Prompt.bmpr.zip](https://github.com/prominic/Moonshine-IDE/files/3098880/2019_04_19__Subproject_Import_Prompt.bmpr.zip)

Here is my proposed algorithm:
```
function void openProject(File directory) {
    if (isProject(directory)) {
        // open project normally
    }
    else {

        Vector subProjects = findSubProjects(directory);
        if (subProjects.size() <= 0) {
            // Go through the project import prompt
        }
        else {
            // prompt the user with the list of subprojects
        }
    }

}

// Check if the project has any Moonshine project configuration files
function boolean isProject(File directory) {
    return hasFile(directory, "*.asproj") || 
           hasFile(directory, "*.javaproj") || 
           hasFile(directory, "*.visualeditorproj");
}

// Iterate through the subdirectories to check for any that are recognized as a project.
function Vector findSubProjects(File baseDirectory) {
    Vector subprojects = new Vector();
    Queue processingQueue = new Queue();

    // Start with the immediate children of the base directory
    processingQueue.pushAll(baseDirectory.getChildren()) 

    while (!processingQueue.isEmpty()) {
        File curDirectory = processingQueue.pop();  // Get and remove the first entry
        if (isProject(curDirectory)) {
            subprojects.add(curDirectory);
            // NOTE:  any subdirectories of this project will *not* be processed
        }
        else {
            // Keep searching the immediate children of this directory
            processingQueue.add(curDirectory.getChildren())
        }
    }

    return subprojects
}
```

**Comment by JoelProminic** [comment]
I talked with @rat-moonshine about this, and we agreed that it was probably better to limit the depnth of the search to keep the performance reasonable.

We need to support a least a depth of 2, for the Moonshine-IDE example above, and for a traditional SVN repository (though this is probably not the best way to checkout an SVN repository):

    https://svn...
    - branches
    - tags
    - trunk
        - Project1
        - Project2
        ...


For now, let's limit the depth to 2.  If we find that we need to expand it in practical tests, we can update this later.

**Comment by rat-moonshine** [comment]
Projects-to-open selection window is now available for both Git and SVN checkouts.

**Comment by rat-moonshine** [comment]
Requirement: We'll probably want to use the same projects-open logic in case of Open/Import Project, too. The updated logic should support 'Open Project with Existing Source' feature.

**Comment by rat-moonshine** [comment]
Open Projects selection popup now also available to Open/Import Project option.

**Comment by JoelProminic** [comment]
When I cloned Moonshine-IDE for #503, I was prompted to open all of the subprojects as expected.  vscode-nextgenas_MoonshineModifications was included in the available project list.  

![image](https://user-images.githubusercontent.com/24532731/56999522-22297a00-6b75-11e9-87e6-0770496776be.png)

However, Moonshine triggered the project import logic instead of opening this project.  The other projects opened properly.

>This directory is missing the Moonshine project configuration files.  Do you wan to generate a new project by locating existing source?


I expected vscode-nextgenas_MoonshineModifications to be ignored if it was not a valid project.  If this is a simple bug, please fix it.  If this is a special case that I was not aware of, we can discuss this tomorrow.

Otherwise, this feature has been working well in my tests.

**Comment by rat-moonshine** [comment]
**vscode-nextgenas_MoonshineModifications** created in Eclipse. Currently, based on Eclipse/Flash-Builder, Moonshine can import a project if the directory contains both `.actionScriptProperties` and `.project` files.

**vscode-nextgenas_MoonshineModifications** directory contains only `.project`; I can make the code opens the project anyway to the sidebar, but there is no existing logic to check through available files and determine if this is a Java project based on particular fields/values. Thus, the project will open as a non-Java project to Moonshine and may not have selective menu options those should available for a Java project.

Currently, for Java-Gradle/Maven projects, Moonshine specifically looks for the existence of `.javaproj` configuration file existence. Then it goes through [this test](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/plugin/java/javaproject/importer/JavaImporter.as#L14) and mark the project as a Java project to avail different type-specific configuration to the project. This existing test doesn't fit against Java project created in Eclipse IDE.

To determine if it's a Java project created in Eclipse, we need to know what property/value we should check in which file, and probably make the said `JavaImporter.test(..)` method more stronger based on that.

**Comment by JoelProminic** [comment]
After I updated the project configuration for vscode-nextgenas_MoonshineModifications, this feature is working as expected.

I noticed a new case with this today:  The [Royale-Examples](https://github.com/prominic/Royale-Examples) project has only one subproject, HelloWorld.  In this case, Moonshine opens the HelloWorld project without prompting the user.  This behavior confused me for a moment (I thought that the repository might have been configured as a normal project), but I don't have any immediate objections to handling it this way.
